src/main/java/com/google/firebase/database/connection/NettyWebSocketClient.java

```diff
@@ -0,0 +1,202 @@
+import io.netty.handler.codec.http.websocketx.WebSocketClientHandshakerFactory;
+import io.netty.handler.codec.http.websocketx.WebSocketFrame;
+import io.netty.handler.codec.http.websocketx.WebSocketHandshakeException;
+import io.netty.handler.codec.http.websocketx.WebSocketVersion;
+import io.netty.handler.codec.http.websocketx.extensions.compression.WebSocketClientCompressionHandler;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.CharsetUtil;
+
+import java.net.URI;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadFactory;
+import javax.net.ssl.SSLException;
+
+/**
+ * A {@link WebsocketConnection.WSClient} implementation based on the Netty framework. Uses
+ * a single-threaded NIO event loop to read and write bytes from a WebSocket connection. Netty
+ * handles all the low-level IO, SSL and WebSocket handshake, and other protocol-specific details.
+ *
+ * <p>This implementation does not initiate connection close on its own. In case of errors or loss
+ * of connectivity, it notifies the higher layer ({@link WebsocketConnection}), which then decides
+ * whether to initiate a connection tear down.
+ */
+class NettyWebSocketClient implements WebsocketConnection.WSClient {
+
+  private static final int DEFAULT_WSS_PORT = 443;
+
+  private final URI uri;
+  private final WebsocketConnection.WSClientEventHandler eventHandler;
+  private final ChannelHandler channelHandler;
+  private final ExecutorService executorService;
+  private final EventLoopGroup group;
+
+  private Channel channel;
+
+  NettyWebSocketClient(
+      URI uri, String userAgent, ThreadFactory threadFactory,
+      WebsocketConnection.WSClientEventHandler eventHandler) {
+    this.uri = checkNotNull(uri, "uri must not be null");
+    this.eventHandler = checkNotNull(eventHandler, "event handler must not be null");
+    this.channelHandler = new WebSocketClientHandler(uri, userAgent, eventHandler);
+    this.executorService = new RevivingScheduledExecutor(
+        threadFactory, "firebase-websocket-worker", GaeThreadFactory.isAvailable());
+    this.group = new NioEventLoopGroup(1, this.executorService);
+  }
+
+  @Override
+  public void connect() {
+    checkState(channel == null, "channel already initialized");
+    try {
+      final SslContext sslContext = SslContextBuilder.forClient()
+          .trustManager(InsecureTrustManagerFactory.INSTANCE).build();
+      Bootstrap bootstrap = new Bootstrap();
+      bootstrap.group(group)
+          .channel(NioSocketChannel.class)
+          .handler(new ChannelInitializer<SocketChannel>() {
+            @Override
+            protected void initChannel(SocketChannel ch) {
+              ChannelPipeline p = ch.pipeline();
+              p.addLast(sslContext.newHandler(ch.alloc(), uri.getHost(), DEFAULT_WSS_PORT));
```

Comment:
`URI.getPort()` returns -1 when the source URI string doesn't explicitly state the port.

Reply:
I am fine with using DEFAULT_WSS_PORT in that case, but if you explicitly specify a port, this code should respect it.  It may not matter much in practice, since prod firebase uses 443, but if somebody on the RTDB team wanted to test against a localhost instance or something, they might need to use a different port.

Reply:
Fair enough. Changed to:

```
int port = uri.getPort() != -1 ? uri.getPort() : DEFAULT_WSS_PORT;
```